### PR TITLE
Add discrimination-ieee754 and proto-lens-discrimination for linear-time sorting.

### PR DESCRIPTION
**Pull request by awpr**
My fix to discrimination finally got merged a few months ago, so we can proceed with this again.

---
This change is [Reviewable](https://reviewable.io/reviews/google/proto-lens/182)

### CODE REVIEW

File: discrimination-ieee754/src/Data/Discrimination/IEEE754.hs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+-- | Discrimination of floating-point numbers via their IEEE754 representation.
+--
+-- This technique is described by http://stereopsis.com/radix.html.
+--
+-- This orders floating-point numbers according to the IEEE754 floating-point
+-- total order.  -NaN < -Infinity < ... < -0 < 0 < ... < +Infinity < +NaN.
+-- NaNs with positive sign are ordered by the value of their payload, and NaNs
+-- with negative sign are reverse-ordered by the value of their payload.
+
+module Data.Discrimination.IEEE754 where
```
Comment (judah) [inline discrimination-ieee754/src/Data/Discrimination/IEEE754.hs]: Use an explicit export list, and perhaps only export a subset of these functions (e.g. just the "group" and "sort" variants)?

Up to you.

File: proto-lens-discrimination/package.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    - Data.ProtoLens.Discrimination
+
+tests:
+
+  discrimination_test:
+    main: disc_test.hs
+    source-dirs: tests
+    dependencies:
+      - HUnit >= 1.3 && < 1.7
+      - proto-lens-arbitrary
```
Comment (judah) [inline proto-lens-discrimination/package.yaml]: Add bounds for proto-lens-arbitrary?

File: proto-lens-discrimination/src/Data/ProtoLens/Discrimination.hs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import Lens.Family2 ((&), (.~), Lens', view)
+
+import Data.ProtoLens.Message
+    ( FieldDescriptor(FieldDescriptor)
+    , FieldTypeDescriptor(..)
+    , FieldAccessor(..)
+    )
+
+-- | Sort values according to a Foldable of field descriptors.
+discFields
```
Comment (judah) [inline proto-lens-discrimination/src/Data/ProtoLens/Discrimination.hs]: You'll probably also want to sort according to the `unknownFields` of the proto, added in #129: https://github.com/google/proto-lens/blob/master/proto-lens/src/Data/ProtoLens/Message.hs#L88

Up to you whether to do that in this PR or just leave it as a TODO.

File: proto-lens-discrimination/src/Data/ProtoLens/Group.hs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.ProtoLens.Group
+    (
+    -- * Comparing Protobuf Messages
+      eqMessage
+
+    -- * 'Group's for Messages
+    , groupingMessage
+    , groupingFields
```
Comment (judah) [inline proto-lens-discrimination/src/Data/ProtoLens/Group.hs]: I suggest putting `groupingFields` and `groupingField` in a separate Haddock section (or the same one as `groupingFieldValue`.  Unlike `groupingMessage`, I'd expect they're pretty unlikely to be used by real-world programs.

(Alternately, just don't export them.)

File: discrimination-ieee754/src/Data/Discrimination/IEEE754.hs (hunk 1 above)
Comment (awpr) [reply discrimination-ieee754/src/Data/Discrimination/IEEE754.hs]: Done.

File: proto-lens-discrimination/src/Data/ProtoLens/Discrimination.hs (hunk 3 above)
Comment (awpr) [reply proto-lens-discrimination/src/Data/ProtoLens/Discrimination.hs]: Added support for unknown fields.  Now 'sortingMessage' considers unknown fields and 'sortingKnownFields' does not.

File: proto-lens-discrimination/package.yaml (hunk 2 above)
Comment (awpr) [reply proto-lens-discrimination/package.yaml]: Done.

File: proto-lens-discrimination/src/Data/ProtoLens/Group.hs (hunk 4 above)
Comment (awpr) [reply proto-lens-discrimination/src/Data/ProtoLens/Group.hs]: Moved them to a subsection.

File: discrimination-ieee754/LICENSE [hunk 5]
```diff
@@ -0,0 +1,30 @@
+Copyright (c) 2016, Google Inc.
```
[the remark this answers is not in the document]
Comment (awpr) [reply discrimination-ieee754/LICENSE]: Done.

File: proto-lens-discrimination/Setup.hs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+-- Copyright 2016 Google Inc. All Rights Reserved.
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
```
[the remark this answers is not in the document]
Comment (awpr) [reply proto-lens-discrimination/Setup.hs]: Done.

**Comment by blackgnezdo** [comment]
@awpr care to rebase and we can merge it then?

**Comment by awpr** [comment]
Rebasing got complicated since it's previously merged in changes from master.  I've merged in master instead, so I'd recommend having GitHub auto-squash this one.

I'm not sure what's going on with Travis, but it looks like the two Cabal-based builds are flaky.  Between https://travis-ci.org/google/proto-lens/builds/379826442?utm_source=github_status&utm_medium=notification and https://travis-ci.org/google/proto-lens/builds/379838964?utm_source=github_status&utm_medium=notification, all the builds have passed at least once, and the failures seem to be unrelated to the code (timeout downloading something and corrupted package installations).

**Comment by blackgnezdo** [comment]
@judah do you have any suggestions about Travis? It seems to be unrelated to code changes. I'm inclined to merge this in unless you disagree.

**Comment by judah** [comment]
Not sure about the Travis issues, but let's not block this PR on them.

**Pull request merged**